### PR TITLE
One-pager for Crossplane Helm Provider

### PR DESCRIPTION
**Pull request by turkenh**
This PR introduces one-pager design document for Crossplane Helm Provider.

[skip ci]

### Description of your changes

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### CODE REVIEW

**Comment by prasek** [comment]
@muvaf @hasheddan can you review when you get a chance so we can move ahead with this into crossplane-contrib? thx!

File: design/one-pager-helm-provider.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+      version: 9.3.19
+    namespace: wordpress
+    values:
+      mariadb:
+        enabled: false
+      externaldb:
+        enabled: true
+    set:
+    - name: wordpressBlogName
+      value: "Hello Crossplane"
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:90]: No difference. If there is no need to parametrize it, then it makes sense to put it under values which should be the case here.

File: design/one-pager-helm-provider.md [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    - name: externalDatabase.host
+      valueFrom:
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:92]: > Does that mean the controller will fetch the given Secret, get the value and provide it as --set argument to helm upgrade at each reconcile?

Yes.

> Does a change in the Secret trigger a new Helm upgrade?

Yes, but I don't think we need to watch secrets/configmaps just for this purpose given we have reconcile period of 2 mins by default, which would make it eventually consistent. Similar to it takes some time (around 2 mins) to have a change in a mounted configmap/secret to be reflected to the filesystem of the pod.

File: design/one-pager-helm-provider.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    as `ResourceExists = false` which will result in `Create` to be called.
+    2. If there is last release, compare last [release](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/release/release.go#L22)
+    information with the desired spec (e.g. `spec.forProvider`)
+        1. If desired state matches with observed, return `ResourceUpToDate = true`
+        2. If desired state differs from observed, return `ResourceUpToDate = false`
+
+[Create](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalClientFns.Create):
+
+1. Fetch the helm chart by constructing URL as  `<spec.forProvider.chart.repository>/<spec.forProvider.chart.name>-<spec.forProvider.chart.version>.tgz`
+2. Prepare desired config by combining `spec.forProvider.values` and `spec.forProvider.set`s
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:134]: Yes, set has precedence, will mention even explicit for helm users :)

File: design/one-pager-helm-provider.md [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+[Update](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalClientFns.Update):
+
+1. Fetch (if not exists already) the helm chart by constructing URL as  `<spec.forProvider.chart.repository>/<spec.forProvider.chart.name>-<spec.forProvider.chart.version>.tgz`
+2. Prepare desired config by combining `spec.forProvider.values` and `spec.forProvider.set`s
+3. Using [upgrade action](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/action/upgrade.go#L71) of
+helm client, `helm upgrade` with the desired config.
+
+#### Namespaced vs Cluster Scoped
+
+Helm releases are namespaced and with Helm 3, Helm itself also keeping release information in the same namespace where
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:147]: Yes, this is something that I was thinking. Probably simply creating the namespace if not exist would be helpful in most cases. I don't see a reason not to do so. I will add to `Create` section above.

File: design/one-pager-helm-provider.md [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+    1. If there is no last release, return [`ExternalObservation`](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalObservation)
+    as `ResourceExists = false` which will result in `Create` to be called.
+    2. If there is last release, compare last [release](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/release/release.go#L22)
+    information with the desired spec (e.g. `spec.forProvider`)
+        1. If desired state matches with observed, return `ResourceUpToDate = true`
+        2. If desired state differs from observed, return `ResourceUpToDate = false`
+
+[Create](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalClientFns.Create):
+
+1. Fetch the helm chart by constructing URL as  `<spec.forProvider.chart.repository>/<spec.forProvider.chart.name>-<spec.forProvider.chart.version>.tgz`
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:133]: Adding link to helm repo url: https://chartmuseum.com/docs/#helm-chart-repository

File: design/one-pager-helm-provider.md [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+            - name: externalDatabase.user
+              valueFrom:
+                secretKeyRef:
+                  key: username
+            - name: externalDatabase.password
+              valueFrom:
+                secretKeyRef:
+                  key: password
+            - name: blogName
+          reclaimPolicy: Delete
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:306]: Is this something each resource controller needs to implement or `providerRef.name` set properly by composition controllers?

File: design/one-pager-helm-provider.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:119]: Added creation of namespace as a step in `Create`

File: design/one-pager-helm-provider.md [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+     this approach.
+    - Allows changing the image used for job, so claims to support helm 3 if you provide an image with helm 3 binary.
+-  https://github.com/rancher/helm-controller
+    - Focuses on k3s and follows the job approach as kubedex/helm-controller
+    - Have couple of important open issues seems related to using jobs
+     (e.g. [#25](https://github.com/rancher/helm-controller/issues/25) and [#21](https://github.com/rancher/helm-controller/issues/21))
+     and does not seem very active
+-  https://github.com/bitnami-labs/helm-crd
+    - Experimental and has been archived.
+
```
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md]: Well, actually this does not invalidate the need to deploying raw Kubernetes YAMLs where it makes sense to have a `provider-kubernetes` at some point, similar to [terraform kubernetes provider](https://www.terraform.io/docs/providers/kubernetes/index.html). So, I don't feel it is an `alternative` to crossplane helm provider.

**Comment by turkenh** [comment]
@muvaf thanks for the review, all of your comments should be responded/resolved now.

File: design/one-pager-helm-provider.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (muvaf) [reply design/one-pager-helm-provider.md:276]: You can implement it manually but there is a set of types & functions that you utilize in the controller-runtime. Here is an example from CloudSQLInstance referring to a Network:
* The types used for `*Ref` and `*Selector` fields: https://github.com/crossplane/provider-gcp/blob/master/apis/database/v1beta1/cloudsql_instance_types.go#L366
* Actual resolver code: https://github.com/crossplane/provider-gcp/blob/master/apis/database/v1beta1/referencers.go#L30

The resolver interface `func (mg *CloudSQLInstance) ResolveReferences(ctx context.Context, c client.Reader) error` is something we expect a resource to implement if they'd like their references to be resolved. In this case it's only 1 field, but you can grow this function as much as you'd like to.

In the managed reconciler, you can see that it's called before any of `ExternalClient` functions: https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go#L520 What the provided ready-made functions do is to check the `Ready` condition of the referenced object and return error if it's not ready so that the reconciler is blocked but you can just copy that function and put your own readiness check as you'd like.

**Comment by kferrone** [comment]
When using Helm, there are many times when a specific Helm chart does not do exactly what is desired. For example, many charts do not open up the `nodeSelector` feature in the values file, therefore you can't set it. Fortunately Helm v3 has a feature called [`--post-renderer`](https://helm.sh/docs/topics/advanced/#post-rendering) which is a flag that accepts a binary. This binary gets the rendered output from Helm as an argument. Then the resources can be modified before getting applied. This feature is extremely useful when you need full control over the output of Helm. 

The other benefit of this approach is keeping Helm charts focused on app logic instead of a proxy to every known feature of K8s through the values file. Then one can patch the chart accordingly upon deployment to fit the specific governance needs. Using kustomize as the post-renderer is truly cool. 😎

Here is my story when a feature is missing. Since I am using FluxCD HelmRelease, and post-renderer is unavailable, I had to completely copy my own version of [BitcoinD](https://github.com/helm/charts/tree/master/stable/bitcoind). This chart is missing a lot of features I was required to set therefore I could not actually use the chart and now I must maintain my own copy on our own helm repo, uggh. 

Another case for me is when there is a bug in a chart, e.g. bitnami/rabbitmq [issue 3094](https://github.com/bitnami/charts/issues/3094) or [issue 2418](https://github.com/bitnami/charts/issues/2418). With post render option, I simply patch the chart and the bug is a non issue. With in cluster Helm like FluxCD, I must get on the github project and post issues and cross fingers the issue is fixed quickly. 

Here are some ideas I've imagined in the CRD for a HelmRelease in cluster with post render options:
 
At very least have a simple patching option allowed. This is the exact interface on [kustomizes patchTransformer](https://kubernetes-sigs.github.io/kustomize/api-reference/kustomization/patches/). I believe as a minimum it could serve most use cases of missing features in a chart without a complex implementation. It would also be a recognized interface. 
```yaml
patches:
- patch: |-
    - op: replace
      path: /some/existing/path
      value: new value
  target:
    kind: MyKind
    labelSelector: "env=dev"
- patch: |-
    - op: add
      path: /spec/template/spec/nodeSelector
      value: 
        node.size: really-big
        aws.az: us-west-2a
  target:
    kind: Deployment
    labelSelector: "env=dev"
```

At very most, it would be very powerful to marry kustomize into the HelmRelease. Then the controller would always have kustomize as a an option to patch the release to your very specific liking. I do not have a very clear idea on how this would look though. 
Best ideas I have had so far:
 - have a configmap/secret with kustomize.yaml and various resources as other keys and select it on the HelmRelease resource. Then controller mounts the secret and simply uses the given kustomization as the patch for the release. Here is how I imagined a kustomization configmap to look. Then each resource in the kustomization can easily be mounted as files in a dir. 

```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: some-kustomz
data:
  kustomization.yaml: |
      apiVersion: kustomize.config.k8s.io/v1beta1
      kind: Kustomization
      commonAnnotations:
        project.author: Cool Team
        project.description: |-
            Big long description for all generated resources from chart. 
      resources:
        - all.yaml # this is the generated resources from helm as a single file, would be required
        - github.com/kubernetes-sigs/kustomize/blob/master/examples/helloWorld # just because
      patches:
          - path: ./patch-deployment.yaml
            target: 
               kind: Deployment
               name: "*"
  patch-deployment.yaml: |
      spec:
        template:
          metadata:
            annotations:
              appmesh.k8s.aws/sidecarInjectorWebhook: disabled
          spec:
            nodeSelector: 
              "eks.amazonaws.com/nodegroup": MyGroup
              "failure-domain.beta.kubernetes.io/zone": us-west-2c
```
Then with the above present, the helm release could simply select it.
```yaml
kustomizeRef:
   name: some-kustomz
```

 - simply point to a remote URL which is a kustomize directory. this is built into kustomize already. This would be ideal for GitOps workflows. 

```yaml
kustomizeRef:
  url: git::ssh://git@example.com/foo/bar
```

Either way, kustomize promotes others to embed kustomize into their own app. This way much less effort is needed in building a patching mechanism. 

For my last note on the idea, a Crossplane HelmRelease should be able to get its `values.yaml` file from a secret or configMap. [This is the current behavior for FluxCD HelmRelease](https://github.com/fluxcd/helm-operator/blob/master/docs/helmrelease-guide/values.md). I would suggest the interface be the same as well for ease of use and transition. 

basically like:
```yaml
  valuesFrom:
  - configMapKeyRef:
      name: default-values
      namespace: my-ns
      key: values.yaml
      optional: false
```

references:
 - https://helm.sh/docs/topics/advanced/
 - https://docs.solo.io/gloo/latest/installation/gateway/kubernetes/helm_advanced/
 - https://github.com/thomastaylor312/advanced-helm-demos/tree/master/post-render

**Comment by prasek** [comment]
Thanks @kferrone! Post-render with kustomize is a great idea and seems to be pretty popular as a way to workaround helm charts that aren't templatized how you need them.

@turkenh Would be great to have support for a kustomize post-render or at least the simple patching option -- would go a long way to making this more usable.
/cc @hasheddan @muvaf

**Comment by lukeweber** [comment]
@kferrone - interesting idea, wonder a little about the lifecycle of KustomizeRef from github as a concept. 

* I suspect a user might want the kustomization reflected on cluster for debugging purposes as a resource.
* Kustomize can rely on local context to drive git auth from a cli, but imagine users might prefer to rely on something like - https://github.com/fluxcd/source-controller that has api support for auth schemes and much more.
* I assume if a configmap or secret changes for the helm-provider it would trigger a re-install, wondering out loud if an in cluster Kustomization change would also trigger a re-install. 

Original background from helm on the post-render proposal for more context - https://github.com/helm/helm/issues/7260

**Comment by turkenh** [comment]
Thanks @kferrone, sounds like a good idea!

Regarding post-renderers in general, I am a bit concerned about having/maintaining separate artifacts (patches, kustomization files) for a helm release instead of having everything packaged nicely in a single helm chart. But I would agree that this could be relatively cleaner than having to maintain own fork and enables focused helm charts instead of bloated ones with every possible configuration.

It makes sense to support this in helm provider considering the flexibility and power this could introduce. We need to ensure we provide a good user experience while porting concepts from cli to controller, which we can achieve by defining some opinionated workflow without losing flexibility as much as possible. 

To start with, I like the patching option which provides the flexibility with a simple, well-defined interface. I would prefer to leave "full kustomize" support as a "future consideration" and not include into initial design to start simple by focusing on the actual thing we need here which is a "helm" controller/provider, e.g. should that `helm+kustomize` controller be still called as "helm" provider? 

However, no matter we chose simple patch or full kustomize option, currently proposed "actual state observation" method (which relies on reading applied chart spec and config back from the cluster) will no longer work and we would rather need to check rendered manifests to decide whether we are in sync or not. I'll work on the design accordingly by also keeping possible future "full kustomize" support in mind.

> For my last note on the idea, a Crossplane HelmRelease should be able to get its values.yaml file from a secret or configMap. This is the current behavior for FluxCD HelmRelease. I would suggest the interface be the same as well for ease of use and transition.

Makes sense, I'll include this in the design doc as well.

File: design/one-pager-helm-provider.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:276]: I see, thanks for the details!

File: design/one-pager-helm-provider.md [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+This provider will enable deployment of helm charts to (remote) Kubernetes Clusters typically provisioned via
+Crossplane. Considering the issues with helm 2 (e.g. security problems regarding tiller and lack of proper go 
+clients/libraries for helm), **we will focus and only support Helm 3**. 
+
+## Design
+
+We will implement a Kubernetes controller watching `Release` custom resources and deploying helm charts with the desired
+configuration. Since this controller needs to interact with Kubernetes API server, it is a good fit for [Kubernetes
+native providers](https://github.com/crossplane/crossplane/blob/master/design/one-pager-k8s-native-providers.md#kubernetes-native-providers)
+concept in Crossplane. By using existing [Kubernetes Provider](https://github.com/crossplane/crossplane/blob/master/design/one-pager-k8s-native-providers.md#proposal-kubernetes-provider-kind)
```
Comment (hasheddan) [inline design/one-pager-helm-provider.md:30]: Glad to see `KubernetesProvider` getting some usage besides just `provider-rook` :) However, I wonder if the machinery around `KubernetesTarget` actually makes more sense here. They accomplish similar goals, but `KubernetesTarget` a little easier to work with as you can reference the managed resource directly and have a secret propagated to your namespace. This, along with some of the other types you mentioned (such as databases, users, roles) seem to be somewhat similar to Terraform provisioners, but in the case of the Helm chart are longer lived (i.e. want to continue to reconcile state).

I think something like the existing `KubernetesTarget` or `KubernetesProvider` is a good starting point, but maybe this Helm provider will be a good test case for what we want a more robust model to look like 👍 

I was also considering if types like `KubernetesProvider` and `KubernetesTarget` should live in core Crossplane. I wonder if they should be eventually instead be included in the "Crossplane client-side" chart we have discussed briefly /cc @muvaf @prasek

File: design/one-pager-helm-provider.md [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+
+1. Using [history action](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/action/history.go#L43)
+of helm client, get last [release](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/release/release.go#L22)
+information.
+    1. If there is no last release, return [`ExternalObservation`](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalObservation)
+    as `ResourceExists = false` which will result in `Create` to be called.
+    2. If there is last release, compare last [release](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/release/release.go#L22)
+    information with the desired spec (e.g. `spec.forProvider`)
+        1. If desired state matches with observed, return `ResourceUpToDate = true`
+        2. If desired state differs from observed, return `ResourceUpToDate = false`
```
Comment (hasheddan) [inline design/one-pager-helm-provider.md:94]: Does this mean if you aren't using the latest version of a helm chart then the status of `Release` resource will always say it is not up to date? It seems like the version specified is the desired version, so if that version is installed correctly then the resource should be up to date. However, I do like the idea of informing a user that the chart is not using latest.

Comment (turkenh) [reply design/one-pager-helm-provider.md:190]: No, we have nothing regarding latest version of the chart here, I guess cause for the confusion is `If there is last release`?

If so, it is the last deployed release of the helm chart and not related to chart version in the chart repository. And yes, as you said, since desired version is specified, if it is installed (with the desired config), we are up to date.

Comment (hasheddan) [reply design/one-pager-helm-provider.md:190]: Understood :+1:

File: design/one-pager-helm-provider.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md:195]: Added `Value Overrides` section: https://github.com/crossplane/crossplane/blob/1fd373f5cbe3efece5946deaeec0b905f4df6ff3/design/one-pager-helm-provider.md#value-overrides

File: design/one-pager-helm-provider.md (hunk 8 above)
Comment (turkenh) [reply design/one-pager-helm-provider.md:30]: Not totally sure about exact the difference between `KubernetesProvider` and `KubernetesTarget`, however, for the POC work and the implementation so far I am using `KubernetesProvider` which looks quite fine. I think propagating connection string to desired namespace will not be a big deal thanks to `writeConnectionSecretsToNamespace` option of composition.

Another concern regarding `KubernetesTarget` is, it looks like a little coupled with `KubernetesApplication` (or workload scheduling in general) which is [planned to be removed](https://github.com/crossplane/crossplane/issues/1595) so I would prefer not to rely on it. This is at least my understanding after skimming through [this](https://github.com/crossplane/crossplane/blob/master/design/one-pager-consuming-k8s-clusters.md) design doc.

However, I'll keep this in mind in case I hit some limitations or issues with `KubernetesProvider`.

File: design/one-pager-helm-provider.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply design/one-pager-helm-provider.md]: Realized that this is not always true and sometimes you may need to append `charts/` after repo URL. Fortunately, helm already has a pull action as well which we can use here instead of dealing with these details. Updated the design doc accordingly.

Good catch, thanks!

File: design/one-pager-helm-provider.md [hunk 10]
```diff
@@ -0,0 +1,524 @@
+# Crossplane Helm Provider
+
+* Owner: Hasan Turken (@turkenh)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
```
Comment (hasheddan) [inline design/one-pager-helm-provider.md:5]:
```suggestion
* Status: Accepted
```

Comment (turkenh) [reply design/one-pager-helm-provider.md]: Done, thanks!

**Comment by hasheddan** [comment]
@turkenh unfortunately the "commit this change" functionality breaks DCO :/

**Comment by negz** [comment]
> should that helm+kustomize controller be still called as "helm" provider?

This is a large part of the reason I've been advocating for a provider that targets a layer above specific deployment tools like Helm or Kustomize. If we wrote a provider for a tool such as [Rancher Fleet](https://github.com/rancher/fleet/tree/master/examples/helm-kustomize) and [ArgoCD](https://github.com/argoproj/argocd-example-apps/tree/master/plugins/kustomized-helm) this problem would be solved for us.

File: design/one-pager-helm-provider.md [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+                secretKeyRef:
+                  key: username
+            - name: externalDatabase.password
+              valueFrom:
+                secretKeyRef:
+                  key: password
+            - name: blogName
+          reclaimPolicy: Delete
+          providerSelector:
+            matchControllerRef: true
```
Comment (negz) [inline design/one-pager-helm-provider.md:379]: I think this is trying to select the Kubernetes provider created in the same composition, right? This functionality (selecting rather than referencing a provider) does not exist today. I'd support introducing it, but I'd prefer to do so for all providers rather than making a special case for the Helm provider.
